custom-transform/app.R

```diff
@@ -0,0 +1,137 @@
+library(teal)
+
+my_transformers <- list(
+  teal_transform_module(
+    label = "Keep first 6 from IRIS",
+    ui = function(id) {
+      ns <- NS(id)
+      div(
+        checkboxInput(ns("check"), label = "Toggle `head(iris)`"),
+      )
+    },
+    server = function(id, data) {
+      moduleServer(id, function(input, output, session) {
+        eventReactive(input$check, {
+          req(data())
+          if (input$check) {
+            within(data(), iris <- head(iris, 6))
+          } else {
+            data()
+          }
+        })
+      })
+    }
+  ),
+  teal_transform_module(
+    label = "Merge datasets to get ANL",
+    ui = function(id) {
+      ns <- NS(id)
+      tagList(
+        div("Choose the two datasets to merge:"),
+        teal.widgets::optionalSelectInput(ns("merge_a"), "Merge A", choices = NULL),
+        teal.widgets::optionalSelectInput(ns("merge_b"), "Merge B", choices = NULL)
+      )
+    },
+    server = function(id, data) {
+      checkmate::assert_class(data, "reactive")
+      moduleServer(id, function(input, output, session) {
+        iv <- shinyvalidate::InputValidator$new()
+        iv$add_rule("merge_a", shinyvalidate::sv_required("Please select dataset A"))
+        iv$add_rule("merge_b", shinyvalidate::sv_required("Please select dataset B"))
+        iv$enable()
+        
+        reactive_datanames <- reactive({
+          req(data())
+          teal.data::datanames(data())
+        })
+        observeEvent(reactive_datanames(), {
+          selected_a <- isolate(input$merge_a)
+          if (identical(selected_a, "")) selected_a <- restoreInput(session$ns("merge_a"), NULL)
+          teal.widgets::updateOptionalSelectInput(
+            session = session,
+            inputId = "merge_a",
+            choices = reactive_datanames(),
+            selected = restoreInput(session$ns("merge_a"), selected_a)
+          )
+
+          selected_b <- isolate(input$merge_b)
+          if (identical(selected_b, "")) selected <- restoreInput(session$ns("merge_b"), NULL)
+          teal.widgets::updateOptionalSelectInput(
+            session = session,
+            inputId = "merge_b",
+            choices = reactive_datanames(),
+            selected = restoreInput(session$ns("merge_b"), selected_b)
+          )
+        })
+        
+        merge_a <- reactive(input$merge_a)
+        merge_b <- reactive(input$merge_b)
+        
+        reactive({
+          new_data <- within(
+            data(),
+            ANL <- dplyr::left_join(merge_a, merge_b),
+            merge_a = tryCatch(as.name(merge_a()), error = function(e) as.name("DatasetA")),
+            merge_b = tryCatch(as.name(merge_b()), error = function(e) as.name("DatasetA"))
```

Comment:
I don't think `tryCatch` should be exposed in "education" materials
```suggestion
          req(merge_a(), merge_b())
          new_data <- within(
            data(),
            ANL <- dplyr::left_join(merge_a, merge_b),
            merge_a = as.name(merge_a()),
            merge_b = as.name(merge_b())
```